statefun-e2e-tests/statefun-smoke-e2e-java/src/main/java/org/apache/flink/statefun/e2e/smoke/CommandInterpreter.java

```diff
@@ -0,0 +1,128 @@
+
+import java.time.Duration;
+import java.util.Objects;
+import org.apache.flink.statefun.e2e.smoke.generated.Command;
+import org.apache.flink.statefun.e2e.smoke.generated.Commands;
+import org.apache.flink.statefun.e2e.smoke.generated.VerificationResult;
+import org.apache.flink.statefun.sdk.java.*;
+import org.apache.flink.statefun.sdk.java.message.EgressMessage;
+import org.apache.flink.statefun.sdk.java.message.EgressMessageBuilder;
+import org.apache.flink.statefun.sdk.java.message.Message;
+import org.apache.flink.statefun.sdk.java.message.MessageBuilder;
+
+public final class CommandInterpreter {
+  private final Ids ids;
+  private static final Duration sendAfterDelay = Duration.ofMillis(1);
+
+  public CommandInterpreter(Ids ids) {
+    this.ids = Objects.requireNonNull(ids);
+  }
+
+  public void interpret(ValueSpec<Long> state, Context context, Message message) {
+    if (message.is(SOURCE_COMMAND_TYPE)) {
+      interpret(state, context, message.as(SOURCE_COMMAND_TYPE).getCommands());
+    } else if (message.is(COMMANDS_TYPE)) {
+      interpret(state, context, message.as(COMMANDS_TYPE));
+    } else {
+      throw new IllegalArgumentException("Unrecognized message type " + message.valueTypeName());
+    }
+  }
+
+  private void interpret(ValueSpec<Long> state, Context context, Commands cmds) {
+    for (Command cmd : cmds.getCommandList()) {
+      if (cmd.hasIncrement()) {
+        modifyState(state, context, cmd.getIncrement());
+      } else if (cmd.hasSend()) {
+        send(state, context, cmd.getSend());
+      } else if (cmd.hasSendAfter()) {
+        sendAfter(state, context, cmd.getSendAfter());
+      } else if (cmd.hasSendEgress()) {
+        sendEgress(state, context, cmd.getSendEgress());
+      } else if (cmd.hasVerify()) {
+        verify(state, context, cmd.getVerify());
+      }
+    }
+  }
+
+  private void verify(
+      ValueSpec<Long> state, @SuppressWarnings("unused") Context context, Command.Verify verify) {
+    AddressScopedStorage storage = context.storage();
+    int selfId = Integer.parseInt(context.self().id());
+    long actual = storage.get(state).orElse(0L);
+    long expected = verify.getExpected();
+    VerificationResult verificationResult =
+        VerificationResult.newBuilder()
+            .setId(selfId)
+            .setActual(actual)
+            .setExpected(expected)
+            .build();
+    EgressMessage egressMessage =
+        EgressMessageBuilder.forEgress(VERIFICATION_EGRESS)
+            .withCustomType(VERIFICATION_RESULT_TYPE, verificationResult)
+            .build();
+    context.send(egressMessage);
+  }
+
+  private void sendEgress(
+      @SuppressWarnings("unused") ValueSpec<Long> state,
+      Context context,
+      @SuppressWarnings("unused") Command.SendEgress sendEgress) {
+    EgressMessage egressMessage =
+        EgressMessageBuilder.forEgress(DISCARD_EGRESS)
+            .withCustomType(COMMAND_TYPE, Command.getDefaultInstance())
+            .build();
```

Comment:
This operation can be simplified by just sending strings, since the message is discarded anyways.
```suggestion
    EgressMessage egressMessage =
        EgressMessageBuilder.forEgress(DISCARD_EGRESS)
            .withValue("discarded-message")
            .build();
```

Reply:
OK.